server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java

```diff
@@ -0,0 +1,409 @@
+            convert();
+        }
+    }
+
+    /**
+     * If we still holding values in a set, convert this filter into an approximate, cuckoo-backed filter.
+     * This will create a list of CuckooFilters, and null out the set of hashes
+     */
+    void convert() {
+        if (isSetMode == false) {
+            throw new IllegalStateException("Cannot convert SetBackedScalingCuckooFilter to approximate " +
+                "when it has already been converted.");
+        }
+        long oldSize = getSizeInBytes();
+
+        filters = new ArrayList<>();
+        CuckooFilter t = new CuckooFilter(capacity, fpp, rng);
+        // Cache the chosen numBuckets for later use
+        numBuckets = t.getNumBuckets();
+        fingerprintMask = t.getFingerprintMask();
+        bitsPerEntry = t.getBitsPerEntry();
+
+        hashes.forEach(t::add);
+        filters.add(t);
+
+        hashes = null;
+        isSetMode = false;
+
+        breaker.accept(-oldSize); // this zeros out the overhead of the set
+        breaker.accept(getSizeInBytes()); // this adds back in the new overhead of the cuckoo filters
+
+    }
+
+    /**
+     * Get the approximate size of this datastructure.  Approximate because only the Set occupants
+     * are tracked, not the overhead of the Set itself.
+     */
+    public long getSizeInBytes() {
+        long bytes = 13; // fpp (double), threshold (int), isSetMode (boolean)
+        if (hashes != null) {
+            bytes = (hashes.size() * 16);
+        }
+        if (filters != null) {
+            bytes += filters.stream().mapToLong(CuckooFilter::getSizeInBytes).sum();
+        }
+        return bytes;
+    }
+
+
+    /**
+     * Merge `other` cuckoo filter into this cuckoo.  After merging, this filter's state will
+     * be the union of the two.  During the merging process, the internal Set may be upgraded
+     * to a cuckoo if it goes over threshold
+     */
+    public void merge(SetBackedScalingCuckooFilter other) {
+        // Some basic sanity checks to make sure we can merge
+        if (this.threshold != other.threshold) {
+            throw new IllegalStateException("Cannot merge other CuckooFilter because thresholds do not match: ["
+                + this.threshold + "] vs [" + other.threshold + "]");
+        }
+        if (this.capacity != other.capacity) {
+            throw new IllegalStateException("Cannot merge other CuckooFilter because capacities do not match: ["
+                + this.capacity + "] vs [" + other.capacity + "]");
+        }
+        if (this.fpp != other.fpp) {
+            throw new IllegalStateException("Cannot merge other CuckooFilter because precisions do not match: ["
+                + this.fpp + "] vs [" + other.fpp + "]");
+        }
+
+        if (isSetMode && other.isSetMode) {
+            // Both in sets, merge collections then see if we need to convert to cuckoo
+            hashes.addAll(other.hashes);
+            maybeConvert();
```

Comment:
Hmm, yeah, we can go about twice over the threshold.  Tricky to estimate if we should convert to a filter first though.  If both sets are duplicates of each other, the total size might not change (or change much).  But we won't know that until we've merged them together.

I think it won't matter too much if we go twice over the threshold, since the threshold is set very low relative to the size of the filters.  E.g. the current (hard coded) threshold is 10,000 hashes.  So 20k longs would be ~ 160kb, compared to the initial filter size of ~1.7mb.

Reply:
that works for me